### PR TITLE
Don't special case NN functions in gen_variable_type.py

### PR DESCRIPTION
**Pull request by colesbury**
This modifies NN binding in ATen so that the xxx_forward functions now
return buffers instead of taking them as inputs. The NN functions with
no suffix are implemented in Type.cpp. They call the xxx_forward
variants and discard any returned buffers.

This simplifies derivatives for NN functions. The derivatives are now
defined on the xxx_forward functions and buffers are treated as any
other input.

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -138,7 +138,7 @@
 - name: cat(TensorList tensors, int64_t dim)
   tensors: cat_tensors_backward(grad, to_arg_sizes(tensors, dim), dim)
 
-- name: cauchy(Tensor self, double median, double sigma, Generator generator)
+- name: cauchy_(Tensor self, double median, double sigma, Generator generator)
```
Comment (gchanan) [inline tools/autograd/derivatives.yaml]: ok, now I'm confused about when I need to write an inplace function name here.  I.e. there is no add_ here, but there is a cauchy_.  Is it if there is _only_ an inplace function?  Below it says "In-place functions wil use the out-of-place derivative definition if there is no in-place specific derivative" -- that's ambiguous about whether I can `cachy(` here or not.  It's an out-of-place definition, but is it valid?  Can you add a comment to the top describing a rule for people to use?

File: aten/src/ATen/function_wrapper.py [hunk 2]
```diff
@@ -489,8 +515,17 @@ def process_option(option, output_options):
         option['method_prefix_derived'] = '' if broadcast_arg is None else 's_'
         env = nested_dict(option, top_env)
 
+        mode = option['mode']
         abstract = True
-        if broadcast_arg is None:
+        if mode == 'NN' and option.get('cimpls') is None:
```
Comment (gchanan) [inline aten/src/ATen/function_wrapper.py]: can you add a comment describing the cases when mode == 'NN' and 'cimpls' is None vs not-None?  I don't quite follow when each code path is triggered.

File: tools/autograd/load_derivatives.py [hunk 3]
```diff
@@ -296,3 +271,25 @@ def all_saved_variables(derivatives, key):
 
 def to_camel_case(name):
     return ''.join([p.title() for p in name.split('_')])
+
+
+def match_declarations_with_autograd_functions(declarations, autograd_functions):
+    """Sets the "derivative" key on declarations to matching autograd functions
+
+    In-place functions wil use the out-of-place derivative definition if there
```
Comment (gchanan) [inline tools/autograd/load_derivatives.py]: nit: will

**Review by gchanan — APPROVED** [review]
lgtm, I think a couple of comments would help clarify things.

**Comment by dzhulgakov** [comment]
Hey folks, please when introducing core changes make sure it doesn't break ONNX export. Every diff in PyTorch should have a connected PR in onnxbot/onnx-fb-universe repo (https://github.com/onnxbot/onnx-fb-universe/pull/145 in this case) and it has the build status. Please kindly check it before accepting the diff.

On the side we should look at how to make it a blocking check explicitly.

**Comment by apaszke** [comment]
We should make onnxbot report it as part of the build status. The PR only shows up as referenced and has only a green “open” label, so it’s easy to miss

**Pull request merged**